File: src/Samples/XAML_Features/MeasureArrange/RadialPanel.fs
namespace OpenSilver.Samples.Showcase

open System
open System.Collections.Generic
open System.Linq
open System.Text
open System.Threading.Tasks
open System.Windows.Controls
open System.Windows

//code taken from: https://jobijoy.blogspot.com/2008/04/simple-radial-panel-for-wpf-and.html
type RadialPanel() =
    inherit Panel()

    // MeasureOverride: Measure each child and give as much room as they want
    override this.MeasureOverride(availableSize: Size) =
        for elem in this.Children do
            // Give Infinite size as the available size for all the children
            elem.Measure(Size(Double.PositiveInfinity, Double.PositiveInfinity))
        base.MeasureOverride(availableSize)

    // ArrangeOverride: Arrange all children based on the geometric equations for the circle
    override this.ArrangeOverride(finalSize: Size) =
        if this.Children.Count = 0 then finalSize
        else
            let mutable angle = 0.0

            // Degrees converted to Radian by multiplying with PI / 180
            let incrementalAngularSpace = (360.0 / float this.Children.Count) * (Math.PI / 180.0)

            // An approximate radii based on the available size
            let radiusX = finalSize.Width / 2.4
            let radiusY = finalSize.Height / 2.4

            for elem in this.Children do
                // Calculate the point on the circle for the element
                let childPoint = Point(Math.Cos(angle) * radiusX, -Math.Sin(angle) * radiusY)

                // Offsetting the point to the Available rectangular area which is finalSize
                let actualChildPoint = Point(finalSize.Width / 2.0 + childPoint.X - elem.DesiredSize.Width / 2.0, finalSize.Height / 2.0 + childPoint.Y - elem.DesiredSize.Height / 2.0)

                // Call Arrange method on the child element by giving the calculated point as the placementPoint
                elem.Arrange(Rect(actualChildPoint.X, actualChildPoint.Y, elem.DesiredSize.Width, elem.DesiredSize.Height))

                // Calculate the new angle for the next element
                angle <- angle + incrementalAngularSpace

            finalSize

File: src/Samples/XAML_Panels/MeasureArrange/RadialPanel.vb
Imports System
Imports System.Windows
Imports System.Windows.Controls

Namespace OpenSilver.Samples.Showcase

    'code inspired from: https://jobijoy.blogspot.com/2008/04/simple-radial-panel-for-wpf-and.html
    Public Class RadialPanel
        Inherits Panel

        'Measure each children and give as much room as they want 
        Protected Overrides Function MeasureOverride(ByVal availableSize As Size) As Size
            For Each elem As UIElement In Children
                'Give Infinite size as the avaiable size for all the children
                elem.Measure(New Size(Double.PositiveInfinity, Double.PositiveInfinity))
            Next

            Return MyBase.MeasureOverride(availableSize)
        End Function

        'Arrange all children based on the geometric equations for the circle.
        Protected Overrides Function ArrangeOverride(ByVal finalSize As Size) As Size
            If Children.Count = 0 Then Return finalSize

            Dim _angle As Double = 0
            'Degrees converted to Radian by multiplying with PI/180
            Dim _incrementalAngularSpace As Double = (360.0 / Children.Count) * (Math.PI / 180)

            'An approximate radii based on the avialable size , obviously a better approach is needed here.
            Dim radiusX As Double = finalSize.Width / 2.4
            Dim radiusY As Double = finalSize.Height / 2.4

            For Each elem As UIElement In Children
                'Calculate the point on the circle for the element
                Dim childPoint As Point = New Point(Math.Cos(_angle) * radiusX, -Math.Sin(_angle) * radiusY)
                'Offsetting the point to the Avalable rectangular area which is FinalSize.
                Dim actualChildPoint As Point = New Point(finalSize.Width / 2 + childPoint.X - elem.DesiredSize.Width / 2, finalSize.Height / 2 + childPoint.Y - elem.DesiredSize.Height / 2)

                'Call Arrange method on the child element by giving the calculated point as the placementPoint.
                elem.Arrange(New Rect(actualChildPoint.X, actualChildPoint.Y, elem.DesiredSize.Width, elem.DesiredSize.Height))

                'Calculate the new _angle for the next element
                _angle += _incrementalAngularSpace
            Next

            Return finalSize
        End Function
    End Class
End Namespace
